### PR TITLE
feat: simplify logic in `propagated` -> `calculate states at`

### PR DESCRIPTION
**Pull request by vishwa2710**
- Simplify the logic in `calculate_states_at` to not account for the various edge cases of duplicate instants, unsorted instants etc.
- Use test class in Propagated.test to reduce the boiler plate code, and remove tests that were for the edge cases mentioned above.

### CODE REVIEW

**Review by Derollez — COMMENTED** [review]
See https://github.com/open-space-collective/open-space-toolkit-astrodynamics/pull/91 and let's discuss the update on `cachedStateArray_`

File: test/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit/Models/Propagated.test.cpp [hunk 1]
```diff
@@ -1327,53 +610,19 @@ TEST (OpenSpaceToolkit_Astrodynamics_Trajectory_Orbit_Models_Propagated, Calcula
...
-        // Satellite dynamics setup
-        SatelliteDynamics satelliteDynamics = { customEnvironment, satelliteSystem } ;
-
-        // Construct default numerical solver
-        NumericalSolver numericalSolver = { NumericalSolver::LogType::NoLog, NumericalSolver::StepperType::RungeKuttaFehlberg78, 5.0, 1.0e-15, 1.0e-15 } ;
-
         // Setup Propagated model and perform propagation
-        const Propagated propagatedModel = {satelliteDynamics, numericalSolver, cachedStateArray} ;
+        const Propagated propagatedModel = {satelliteDynamics, defaultnumericalSolver_, cachedStateArray} ;
         const Array<State> propagatedStateArray = propagatedModel.calculateStatesAt(instantArray) ;
```
[the remark this answers is not in the document]
Comment (Derollez) [reply test/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit/Models/Propagated.test.cpp:624]: See !91

File: src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit/Models/Propagated.cpp [hunk 2]
```diff
@@ -158,151 +158,115 @@ Array<State>                    Propagated::calculateStatesAt               (
...
+        Array<State> averagedStates = Array<State>::Empty() ;
+        averagedStates.reserve(instants.getSize()) ;
+
+        for (Size k = 0 ; k < instants.getSize() ; ++k)
         {
 
-            // Call numerical solver to perform propagation
-            Array<State> propagatedPositiveStateVectorArray = propagator_.calculateStatesAt(startState, desiredPositiveInstantArray) ;
-            sortedStateArray.add(propagatedPositiveStateVectorArray) ;
+            Real forwardWeight = (instants[k] - this->cachedStateArray_[i].getInstant()).inSeconds() / durationBetweenStates ;
```
[the remark this answers is not in the document]
Comment (vishwa2710) [reply src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit/Models/Propagated.cpp:241]: Yup we should have no duplicated 👍

File: src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit/Models/Propagated.cpp [hunk 3]
```diff
@@ -158,151 +158,115 @@ Array<State>                    Propagated::calculateStatesAt               (
...
-        unsortedStateArray[instantArrayPairs[k].second] = sortedStateArray[k] ;
+        instants.add(anInstantArray[j]) ;
     }
 
-    // Add newly propagated states to the cachedStateArray and then resort
-    cachedStateArray_.add(sortedStateArray) ;
-
-    // Re-sort and sanitize cachedStateArray of duplicates and duplicate instants that have different positions/velocities
-    sanitizeCachedArray() ;
+    allStates.add(propagator_.calculateStatesAt(this->cachedStateArray_.accessLast(), instants)) ;
```
[the remark this answers is not in the document]
Comment (vishwa2710) [reply src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit/Models/Propagated.cpp:267]: Yeah we don't update the local cached state array. I don't think we should necessarily update the local cached state array, I think it can eventually become really large. And then we have to get into memory management (based on the use cases). If the user wants to update it, we can provide a setter. That's a better path. Which then also `sanitizes` so we're protected. That way the user can:
- take a cached state array
- propgate to some instants, get a new state array
- join the two and set them if they want
- propagate again
If we find in the future this is a pattern we find ourselves repeating a lot then we can include it in `calculateStatesAt`. But I don't foresee this happening.

[the remark this answers is not in the document]
Comment (Derollez) [reply src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit/Models/Propagated.cpp:267]: > If the user wants to update it, we can provide a setter
Okay, I like that! At least we can play with it and use it and that will give us a good appreciation of the point we make here:
> get into memory management (based on the use cases)
But that way we have a possibility (on user side) to do both (update, not update)

**Review by Derollez — APPROVED** [review]
